core/src/main/java/com/graphhopper/routing/weighting/custom/PriorityCustomConfig.java

```diff
@@ -0,0 +1,134 @@
+                    priorityList.add(new BooleanToValueEntry(encodedValue, ((Number) value).doubleValue(), 1));
+                }
+            } else if (value instanceof Map) {
+                EnumEncodedValue enumEncodedValue = getEV(lookup, "priority", key, EnumEncodedValue.class);
+                Class<? extends Enum> enumClass = factory.findValues(key);
+                double[] values = Helper.createEnumToDoubleArray("priority." + key, 1, 0, 100,
+                        enumClass, (Map<String, Object>) value);
+                normalizeFactor(values, 1);
+                priorityList.add(new EnumToValueEntry(enumEncodedValue, values));
+            } else {
+                throw new IllegalArgumentException("Type " + value.getClass() + " is not supported for " + key + " in 'priority'");
+            }
+        }
+    }
+
+    static <T extends EncodedValue> T getEV(EncodedValueLookup lookup, String name, String key, Class<T> encodedValueType) {
+        if (!lookup.hasEncodedValue(key))
+            throw new IllegalArgumentException("Cannot find '" + key + "' specified in '" + name + "'");
+        return lookup.getEncodedValue(key, encodedValueType);
+    }
+
+    /**
+     * Pick the maximum value and if it is greater than the specified max we divide all values with it - i.e. normalize it.
+     * <p>
+     * The purpose of this method is to avoid factors above max which makes e.g. Weighting.getMinWeight simpler (as maximum priority is 1)
+     * and also ensures that the landmark algorithm still works (weight can only increased without effecting optimality).
+     */
+    private static void normalizeFactor(double[] values, final double max) {
+        double tmpMax = max;
+        for (int i = 0; i < values.length; i++) {
+            tmpMax = Math.max(tmpMax, values[i]);
+        }
+        if (tmpMax > max)
+            for (int i = 0; i < values.length; i++) {
+                values[i] = values[i] / tmpMax;
+            }
+    }
+
+    void add(EncodedValueLookup lookup, Double value, String name, String encValue) {
+        if (value == null)
+            return;
+        if (!lookup.hasEncodedValue(encValue))
+            throw new IllegalArgumentException("You cannot use " + name + " as encoded value '" + encValue + "' was not enabled on the server.");
+        priorityList.add(new MaxValueEntry(name, lookup.getDecimalEncodedValue(encValue), value));
+    }
+
+    /**
+     * @return weight without unit. The lower the priority is the higher the weight of the specified edge will be.
+     */
+    public double calcPriority(EdgeIteratorState edge, boolean reverse) {
+        double priority = 1;
+        for (int i = 0; i < priorityList.size(); i++) {
+            EdgeToValueEntry entry = priorityList.get(i);
+            double value = entry.getValue(edge, reverse);
+            priority *= value;
+            if (priority <= 0)
```

Comment:
Why do we return a potentially negative priority here? What if `entry.getValue()` is trying to say: I dont care about this edge give me the default? Should it not rather be something like:

```java
double value = entry.getValue(edge, reverse);
if (value == SOME_SPECIAL_VALUE)
    continue;
```

? The `SOME_SPECIAL_VALUE` could be maybe `-1` or something, but in `getValue()` it should be clear what we can return to indicate a missing value.

Reply:
This is indeed very misleading. I want to return early in case there is a value equal to 0. And negative values should be excluded earlier so we should be able to replace this with `==`.

As I'm not even sure if this is of any value (performance-wise) we should probably remove it completely?

Reply:
Is it not like we traverse the whole priorityList and multiply all the values (0.5,0.5)->priority=0.25. And if one of the values indicates 'I do not care' we just skip it (0.5,-1,0.5)->priority=0.25? Why do we stop the first time we encounter one of the 'do not care' values? When I am implementing `EdgeToValueEntry#getValue` I want to be able to either return a value that is used just like I return it **or** return some special value that means 'I do not care' for this entry?

Reply:
There was once a "stop early" criterion but this was for a no longer existing entry called "average_speed". And as soon as the first encoded value matched it did not look further. But I replaced this with "max_speed" and there is no longer a need for the also rather unintuitive "stop early" thing.

The case here is really only meant to return as soon as one value is 0. To make this more clear I should have returned 0.

> When I am implementing EdgeToValueEntry#getValue I want to be able to either return a value that is used just like I return it or return some special value that means 'I do not care' for this entry?

Yes, this can be done with default values like 1 for priority or "default" for SpeedCustomConfig (see the parameter in Helper.createEnumToDoubleArray)

Reply:
https://github.com/graphhopper/graphhopper/pull/1841/commits/f071db27cf4723435e811be94d664dca64d6abde